File: backends/java/zappy-domain/src/test/java/com/zappymart/domain/ordering/OrderTest.java
package com.zappymart.domain.ordering;

import com.zappymart.domain.builders.CartBuilder;
import com.zappymart.domain.cart.Cart;
import com.zappymart.domain.catalogue.Product;
import com.zappymart.domain.promotions.PromotionRule;
import com.zappymart.domain.shared.Money;
import com.zappymart.domain.shared.Result;
import com.zappymart.domain.shared.UserError;
import com.zappymart.domain.shared.UserErrorCode;
import org.junit.jupiter.api.Test;

import java.time.Instant;

import static com.zappymart.domain.builders.CartBuilder.aCart;
import static com.zappymart.domain.builders.ProductBuilder.aProduct;
import static org.assertj.core.api.Assertions.assertThat;

class OrderTest {

    private static final Instant NOW = CartBuilder.A_MOMENT;

    private static final Product SHIRT = aProduct().withId("product-18").named("MBJ Boat Neck")
            .costing(985).withStock(25).build();

    @Test
    void placesAnOrderThatKeepsTheNamesAndPricesOfTheMoment() {
        Cart cart = aCart().ownedBy("customer-01").holding(SHIRT, 2).build();

        Order order = Order.place("order-01", "ZM-000001", "customer-01", cart, NOW).valueOrThrow();

        assertThat(order.status()).isEqualTo(OrderStatus.PAID);
        assertThat(order.lines()).singleElement().satisfies(line -> {
            assertThat(line.productName()).isEqualTo("MBJ Boat Neck");
            assertThat(line.unitPrice()).isEqualTo(Money.euro(985));
            assertThat(line.lineTotal()).isEqualTo(Money.euro(1970));
        });
        assertThat(order.subtotal()).isEqualTo(Money.euro(1970));
        assertThat(order.shipping()).isEqualTo(Money.euro(495));
        assertThat(order.discount()).isEqualTo(Money.zero());
        assertThat(order.total()).isEqualTo(Money.euro(2465));
        assertThat(order.promotionCode()).isNull();
        assertThat(order.placedAt()).isEqualTo(NOW);
    }

    @Test
    void copiesThePromotionCodeAsText() {
        Cart cart = aCart().ownedBy("customer-01").holding(SHIRT, 2)
                .withPromotion("FREESHIP", new PromotionRule.FreeShipping()).build();

        Order order = Order.place("order-01", "ZM-000001", "customer-01", cart, NOW).valueOrThrow();

        assertThat(order.promotionCode()).isEqualTo("FREESHIP");
        assertThat(order.shipping()).isEqualTo(Money.zero());
        assertThat(order.total()).isEqualTo(Money.euro(1970));
    }

    @Test
    void refusesAnEmptyCart() {
        Result<Order> result = Order.place("order-01", "ZM-000001", "customer-01", aCart().build(), NOW);

        assertThat(reasonOf(result)).isEqualTo(UserErrorCode.CART_EMPTY);
    }

    @Test
    void refusesWhenOneLineHasOutrunItsStock() {
        Product lastOne = aProduct().withId("product-12").named("WD 4TB Gaming Drive").withStock(1).build();
        Cart cart = aCart().ownedBy("customer-01").holding(SHIRT, 1).holding(lastOne, 2).build();

        Result<Order> result = Order.place("order-01", "ZM-000001", "customer-01", cart, NOW);

        assertThat(reasonOf(result)).isEqualTo(UserErrorCode.OUT_OF_STOCK);
        assertThat(result.errors().getFirst().message()).contains("WD 4TB Gaming Drive");
    }

    @Test
    void carriesItsPlacementEvent() {
        Cart cart = aCart().ownedBy("customer-01").holding(SHIRT, 2)
                .withPromotion("WELCOME10", new PromotionRule.PercentageOff(10)).build();

        OrderPlaced event = Order.place("order-01", "ZM-000001", "customer-01", cart, NOW)
                .valueOrThrow().placementEvent();

        assertThat(event.orderId()).isEqualTo("order-01");
        assertThat(event.orderNumber()).isEqualTo("ZM-000001");
        assertThat(event.customerId()).isEqualTo("customer-01");
        assertThat(event.promotionCode()).isEqualTo("WELCOME10");
        assertThat(event.occurredAt()).isEqualTo(NOW);
    }

    private static UserErrorCode reasonOf(Result<Order> result) {
        return result.errors().stream().map(UserError::code).findFirst().orElseThrow();
    }
}

File: backends/kotlin/zappy-domain/src/test/kotlin/nl/zappymart/domain/ordering/OrderTest.kt
package nl.zappymart.domain.ordering

import nl.zappymart.domain.builders.A_MOMENT
import nl.zappymart.domain.builders.aCart
import nl.zappymart.domain.builders.aProduct
import nl.zappymart.domain.shared.Money
import nl.zappymart.domain.shared.Result
import nl.zappymart.domain.shared.UserErrorCode
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class OrderTest {

    private val jacket = aProduct().withId("product-03").named("Cotton jacket").costing(5599).withStock(8).build()

    @Test
    fun `an order keeps the names, the prices and the totals of the moment`() {
        val cart = aCart().holding(jacket, 2).withPercentageOff("WELCOME10", 1120).build()
        val placed = Order.place(cart, "customer-01", "order-01", "ZM-1", A_MOMENT) as Result.Success
        val order = placed.value
        assertThat(order.status).isEqualTo(OrderStatus.PAID)
        assertThat(order.lines).singleElement().satisfies({ line ->
            assertThat(line.productName).isEqualTo("Cotton jacket")
            assertThat(line.unitPrice).isEqualTo(Money.euro(5599))
            assertThat(line.lineTotal).isEqualTo(Money.euro(11198))
        })
        assertThat(order.subtotal).isEqualTo(Money.euro(11198))
        assertThat(order.discount).isEqualTo(Money.euro(1120))
        assertThat(order.shipping).isEqualTo(Money.NOTHING)
        assertThat(order.total).isEqualTo(Money.euro(10078))
        assertThat(order.promotionCode?.value).isEqualTo("WELCOME10")
    }

    @Test
    fun `an empty cart cannot be ordered`() {
        val refused = Order.place(aCart().build(), "customer-01", "order-01", "ZM-1", A_MOMENT) as Result.Refused
        assertThat(refused.errors.map { error -> error.code }).containsExactly(UserErrorCode.CART_EMPTY)
    }

    @Test
    fun `a line above the stock stops the whole order and names the product`() {
        val lastOne = aProduct().withId("product-12").named("Gaming drive").withStock(1).build()
        val cart = aCart().holding(jacket).holding(lastOne, 2).build()
        val refused = Order.place(cart, "customer-01", "order-01", "ZM-1", A_MOMENT) as Result.Refused
        assertThat(refused.errors.map { error -> error.code }).containsExactly(UserErrorCode.OUT_OF_STOCK)
        assertThat(refused.errors.first().message).contains("Gaming drive")
    }

    @Test
    fun `a placed order announces itself`() {
        val cart = aCart().holding(jacket).build()
        val order = (Order.place(cart, "customer-01", "order-01", "ZM-1", A_MOMENT) as Result.Success).value
        assertThat(order.placed()).isEqualTo(OrderPlaced("order-01", "ZM-1", "customer-01", null, A_MOMENT))
    }
}
